### ISSUE TITLE
Geopackage in WAL mode freezes QGIS

### ISSUE THREAD

**Issue by dingausmwald** [qgis/QGIS#37623]
Strange problem here, for months. No workaround, just a few hints and fellow sufferes like Bug report #18445 on qgis bug tracker.

Me:
Qgis 3.4.15 (and others, 3.10.6 f.e)
Windows 10
Network drive
Geopackage

When a geopackage is in wal mode on network drive, all qgis instances where this file is opened cant be closed. Even killing qgis via taskmanager wont work.

I have to close all qgis instances/restart windows (of colleagues), export the geopackage from qgis to another location, close qgis (it closes when just one instance is open), and overwrite the old file via filesystem.

Strange: Ive red wal mode is for multiuser stuff. But it only appears when geopackage is edited local. No wal when edited on network drive.

When a colleague edits a geopackage localy, copies it to the network drive, it stays in wal mode, and things start to happen.

More strangely, after 2 days off, i came back to the office, half of the geopackages turned into wal mode on network drive. Colleague told me he coulndt wait for qgis to close and just turned off his mashine, around this time all wal files where created.

What is this wal mode madness? Was so happy to find a successor for shapefiles, but it got worse than before.

**Comment by gioman** [comment]
@dingausmwald you can disable WAL and see if it helps. In the QGIS python console:

`QSettings().setValue("/qgis/walForSqlite3", False)`

**Comment by dingausmwald** [comment]
Thank you. But is this intended behaviour? Are there changes going on?
We changed to geopackage a year ago, and this wal thing is 2020 behaviour. I dont want to change parameters for every system here, just to run in new problems because things will change in coming months. Could somebody explain why this happens? Are we doing something wrong?

Especially the "wal mode is for multiuser, but is deactivated in multiuser environments" (network drive) -thing gets me confused. Can we work with multiple users or will there be data loss?

**Comment by m-kuhn** [comment]
Can you check with QGIS 3.14?

A fix has been implemented for this problem and if we have good test results a backport to 3.10 could be envisioned.

**Comment by dingausmwald** [comment]
File locking on local drives? Like rain underwater. Wal seems turned of on 3.4.15 alrdy. At least there are no wal and shm files on the network drive (normally - let's say for newly created geopackages). 
So I got this right: wal is for file locking, happening when more than 1 person is writing to the same file. Just happens on network drives(correct me if I'm wrong). But it's turned off here because of some limitations of wal I've red. It's turned on for local drives where there's no use for it because there is always just one person writing on local drives. Could you link the discussion? And does smb know since when this has changed? I can remember times wal files where present on network drives and everything was fine (probably bypassing the wal on network drive problem without knowing)

**Comment by m-kuhn** [comment]
File locking is a very basic mechanism avoiding that multiple processes interfere while reading and writing files in parallel on local devices.

More information about wal can be found here: https://www.sqlite.org/wal.html

> All processes using a database must be on the same host computer; WAL does not work over a network filesystem.

There was code in earlier versions that tried to avoid enabling WAL on network devices but it wasn't working correctly and [has only been fixed for QGIS 3.14](https://github.com/qgis/QGIS/pull/36741).

**Comment by dingausmwald** [comment]
QSettings().setValue("/qgis/walForSqlite3", False) does nothing. Implemented it on all 10 windows pcs working with qgis. Nothing changed.

**Comment by dingausmwald** [comment]
Little Feedback:

I finally solved the problem: It all happend when one of our employees left the network with his laptop still running qgis. Gpkg wal mode hasnt been closed properly. Everytime this laptop started qgis in our network, the wal files where created again. The other qgis installations where kind of blocked, working in the network became a pain, delayed file readings, qgis hanging all the time, qgis cant be closed without restarting the whole computer. Installing 3.14.15 on the systems not responsible for the problem didnt help, they still where affected by the wal lock. In addition, they held the wal files, even when the responsible computer was closed already, what made the identification of the problem worse. Installing qgis on the responsible system did the trick. After closing all qgis instances on all systems, restarting the project with 3.14.15 on the responsible windows pc, wal files didnt reappear.

So, with 3.14.15, wal mode seems finaly turned off in network environments. The impairment, that wal mode locks qgis, still seems present, but is no real problem anymore for us.
The walforsqlite command had to be removed manually by us, dont know if this affected the behaviour in any manner. 
Thank you.
